ChattoAdditions/Source/Input/Media/Camera/MediaInputCameraPicker.swift

```diff
@@ -0,0 +1,100 @@
+ AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ THE SOFTWARE.
+*/
+
+import UIKit
+
+protocol MediaInputCameraPickerProtocol {
+    typealias OnImageTakenBlock = (UIImage?) -> Void
+    typealias OnVideoTakenBlock = (URL?) -> Void
+    typealias OnCameraPickerDismissedBlock = () -> Void
+
+    func presentCameraPicker(onImageTaken: @escaping OnImageTakenBlock,
+                             onVideoTaken: @escaping OnVideoTakenBlock,
+                             onCameraPickerDismissed: @escaping OnCameraPickerDismissedBlock)
+}
+
+final class MediaInputCameraPicker: MediaInputCameraPickerProtocol, MediaPickerDelegate {
+    private let mediaPickerFactory: MediaPickerFactory
+    private let presentingControllerProvider: () -> UIViewController?
+    private var MediaPicker: MediaPicker?
+
+    private var onImageTaken: OnImageTakenBlock?
+    private var onVideoTaken: OnVideoTakenBlock?
+    private var onCameraPickerDismissed: OnCameraPickerDismissedBlock?
+
+    init(mediaPickerFactory: MediaPickerFactory,
+         presentingControllerProvider: @escaping () -> UIViewController?) {
+        self.mediaPickerFactory = mediaPickerFactory
+        self.presentingControllerProvider = presentingControllerProvider
+    }
+
+    func presentCameraPicker(onImageTaken: @escaping OnImageTakenBlock,
+                             onVideoTaken: @escaping OnVideoTakenBlock,
+                             onCameraPickerDismissed: @escaping OnCameraPickerDismissedBlock) {
+        guard let presentingController = self.presentingControllerProvider(),
+            let MediaPicker = self.mediaPickerFactory.makeImagePicker(delegate: self) else {
+                onImageTaken(nil)
+                onCameraPickerDismissed()
+                return
+        }
+        self.onImageTaken = onImageTaken
+        self.onVideoTaken = onVideoTaken
+        self.onCameraPickerDismissed = onCameraPickerDismissed
+        self.MediaPicker = MediaPicker
+        presentingController.present(MediaPicker.controller, animated: true, completion: nil)
+    }
+
+    // MARK: - MediaPickerDelegate
+
+    func imagePickerDidFinish(_ picker: MediaPicker, mediaInfo: [UIImagePickerController.InfoKey: Any]) {
+        let mediaType = mediaInfo[UIImagePickerController.InfoKey.mediaType] as? String
+        if let image = mediaInfo[UIImagePickerController.InfoKey.originalImage] as? UIImage,
+            mediaType == InputMediaType.image.UTI {
+            self.finishPickingImage(image, fromPicker: picker.controller)
+        } else if let videoURL = mediaInfo[UIImagePickerController.InfoKey.mediaURL] as? URL,
+            mediaType == InputMediaType.video.UTI {
+            self.finishPickingVideo(videoURL, fromPicker: picker.controller)
+        } else {
+            self.imagePickerDidCancel(picker)
+        }
```

Comment:
It would be nice to clean all variables that stores closures when picker finish its job

Reply:
added cleanup